From 1ce9436ea6c691da0bfe3df861c24afac35f628b Mon Sep 17 00:00:00 2001
From: Huang Pei <huangpei@loongson.cn>
Date: Tue, 26 Jul 2016 09:55:53 +0800
Subject: [PATCH 099/130] fix unaligned gslwxc1/gsldxc1/gssdxc1/gsswxc1
 emulation

Once current process hold the fpu, the following fp access is NOT
going to trigger the CU1 exception, so we need to save the fp cxt;

Change-Id: Iab73bcff0f5e3475f9b0d6245c06463aa58af10b
---
 arch/mips/kernel/unaligned.c | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/arch/mips/kernel/unaligned.c b/arch/mips/kernel/unaligned.c
index ad79ca59..8d58fbf4 100644
--- a/arch/mips/kernel/unaligned.c
+++ b/arch/mips/kernel/unaligned.c
@@ -1335,12 +1335,15 @@ static void emulate_load_store_insn(struct pt_regs *regs,
 			BUG_ON(!used_math());
 			if (!access_ok(VERIFY_READ, addr, 4))
 				goto sigbus;
+			lose_fpu(1);
 
 			LoadW(addr, value, res);
 			if (res)
 				goto fault;
-			compute_return_epc(regs);
 			set_fpr64(current->thread.fpu.fpr, insn.loongson3_lsdc2_format.rt, value);
+			compute_return_epc(regs);
+			own_fpu(1);
+
 			break;
 
 		case 0x7:
@@ -1348,12 +1351,14 @@ static void emulate_load_store_insn(struct pt_regs *regs,
 			BUG_ON(!used_math());
 			if (!access_ok(VERIFY_READ, addr, 8))
 				goto sigbus;
+			lose_fpu(1);
 
 			LoadDW(addr, value, res);
 			if (res)
 				goto fault;
-			compute_return_epc(regs);
 			set_fpr64(current->thread.fpu.fpr, insn.loongson3_lsdc2_format.rt, value);
+			compute_return_epc(regs);
+			own_fpu(1);
 			break;
 
 		}
@@ -1414,13 +1419,15 @@ static void emulate_load_store_insn(struct pt_regs *regs,
 			if (!access_ok(VERIFY_WRITE, addr, 4))
 				goto sigbus;
 
-			compute_return_epc(regs);
+			lose_fpu(1);
 			value = get_fpr64(current->thread.fpu.fpr, insn.loongson3_lsdc2_format.rt);
 
 			StoreW(addr, value, res);
 
 			if (res)
 				goto fault;
+			compute_return_epc(regs);
+			own_fpu(1);
 			break;
 
 
@@ -1431,18 +1438,22 @@ static void emulate_load_store_insn(struct pt_regs *regs,
 			if (!access_ok(VERIFY_WRITE, addr, 8))
 				goto sigbus;
 
-			compute_return_epc(regs);
+			lose_fpu(1);
 			value = get_fpr64(current->thread.fpu.fpr, insn.loongson3_lsdc2_format.rt);
 
 			StoreDW(addr, value, res);
 
 			if (res)
 				goto fault;
+			compute_return_epc(regs);
+			own_fpu(1);
 			break;
 
 
 		}
 		break;
+
+
 #else
 #ifndef CONFIG_CPU_MIPSR6
 	/*
-- 
2.11.0

